In [43]:
!cat game_log_fields.txt

Field(s)  Meaning
    1     Date in the form "yyyymmdd"
    2     Number of game:
             "0" -- a single game
             "1" -- the first game of a double (or triple) header
                    including seperate admission doubleheaders
             "2" -- the second game of a double (or triple) header
                    including seperate admission doubleheaders
             "3" -- the third game of a triple-header
             "A" -- the first game of a double-header involving 3 teams
             "B" -- the second game of a double-header involving 3 teams
    3     Day of week  ("Sun","Mon","Tue","Wed","Thu","Fri","Sat")
  4-5     Visiting team and league
    6     Visiting team game number
          For this and the home team game number, ties are counted as
          games and suspended games are counted from the starting
          rather than the ending date.
  7-8     Home team and league
    9     Home team game number
10-11     Visiting and home team score (unquoted)


In [44]:
import pandas as pd

pd.set_option('max_columns',180)
pd.set_option('max_rows',200000)
pd.set_option('max_colwidth',5000)

In [45]:
game_log = pd.read_csv('game_log.csv')
park_codes = pd.read_csv('park_codes.csv')
person_codes = pd.read_csv('person_codes.csv')
team_codes = pd.read_csv('team_codes.csv')

/dataquest/system/env/python3/lib/python3.4/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (12,13,14,15,19,20,81,82,83,84,85,86,87,88,93,94,95,96,97,98,99,100,105,106,108,109,111,112,114,115,117,118,120,121,123,124,126,127,129,130,132,133,135,136,138,139,141,142,144,145,147,148,150,151,153,154,156,157,160) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [46]:
print("game_log", game_log.shape)
print("park_codes", park_codes.shape)
print("person_codes", person_codes.shape)
print("team_codes", team_codes.shape)

game_log (171907, 161)
park_codes (252, 9)
person_codes (20494, 7)
team_codes (150, 8)


In [47]:
game_log.head().append(game_log.tail())

date  number_of_game day_of_week v_name v_league  v_game_number  \
0       18710504               0         Thu    CL1      NaN              1   
1       18710505               0         Fri    BS1      NaN              1   
2       18710506               0         Sat    CL1      NaN              2   
3       18710508               0         Mon    CL1      NaN              3   
4       18710509               0         Tue    BS1      NaN              2   
171902  20161002               0         Sun    MIL       NL            162   
171903  20161002               0         Sun    NYN       NL            162   
171904  20161002               0         Sun    LAN       NL            162   
171905  20161002               0         Sun    PIT       NL            162   
171906  20161002               0         Sun    MIA       NL            161   

       h_name h_league  h_game_number  v_score  h_score  length_outs  \
0         FW1      NaN              1        0        2         54.0   
1         WS3      NaN              1       20       18         54.0   
2         RC1      NaN              1       12        4         54.0   
3         CH1      NaN              1       12       14         54.0   
4         TRO      NaN              1        9        5         54.0   
171902    COL       NL            162        6        4         60.0   
171903    PHI       NL            162        2        5         51.0   
171904    SFN       NL            162        1        7         51.0   
171905    SLN       NL            162        4       10         51.0   
171906    WAS       NL            162        7       10         51.0   

       day_night completion forefeit protest park_id  attendance  \
0              D        NaN      NaN     NaN   FOR01       200.0   
1              D        NaN      NaN     NaN   WAS01      5000.0   
2              D        NaN      NaN     NaN   RCK01      1000.0   
3              D        NaN      NaN     NaN   CHI01      5000.0   
4              D        NaN      NaN     NaN   TRO01      3250.0   
171902         D        NaN      NaN     NaN   DEN02     27762.0   
171903         D        NaN      NaN     NaN   PHI13     36935.0   
171904         D        NaN      NaN     NaN   SFO03     41445.0   
171905         D        NaN      NaN     NaN   STL10     44615.0   
171906         D        NaN      NaN     NaN   WAS11     28730.0   

        length_minutes v_line_score h_line_score  v_at_bats  v_hits  \
0                120.0    000000000    010010000       30.0     4.0   
1                145.0    107000435    640113030       41.0    13.0   
2                140.0    610020003    010020100       49.0    11.0   
3                150.0    101403111    077000000       46.0    15.0   
4                145.0    000002232    101003000       46.0    17.0   
171902           203.0   0200000202   1100100010       39.0    10.0   
171903           159.0    000001100    00100031x       33.0     8.0   
171904           184.0    000100000    23000002x       30.0     4.0   
171905           192.0    000020200    00100360x       35.0     9.0   
171906           216.0    000230020    03023002x       38.0    14.0   

        v_doubles  v_triples  v_homeruns  v_rbi  v_sacrifice_hits  \
0             1.0        0.0         0.0    0.0               0.0   
1             1.0        2.0         0.0   13.0               0.0   
2             1.0        1.0         0.0    8.0               0.0   
3             2.0        1.0         2.0   10.0               0.0   
4             4.0        1.0         0.0    6.0               0.0   
171902        4.0        1.0         2.0    6.0               0.0   
171903        3.0        0.0         0.0    2.0               0.0   
171904        0.0        0.0         0.0    1.0               0.0   
171905        0.0        0.0         1.0    4.0               0.0   
171906        1.0        1.0         2.0    7.0               1.0   

        v_sacrifice_flies  v_hit_by_pitch  v_walks  v_inte

In [48]:
park_codes.head().append(park_codes.tail())

park_id                                 name  \
0     ALB01                       Riverside Park   
1     ALT01                        Columbia Park   
2     ANA01             Angel Stadium of Anaheim   
3     ARL01                    Arlington Stadium   
4     ARL02        Rangers Ballpark in Arlington   
247   WIL02        BB&T Ballpark at Bowman Field   
248   WNY01     West New York Field Club Grounds   
249   WOR01   Agricultural County Fair Grounds I   
250   WOR02  Agricultural County Fair Grounds II   
251   WOR03       Worcester Driving Park Grounds   

                                          aka           city state  \
0                                         NaN         Albany    NY   
1                                         NaN        Altoona    PA   
2               Edison Field; Anaheim Stadium        Anaheim    CA   
3                                         NaN      Arlington    TX   
4    The Ballpark in Arlington; Ameriquest Fl      Arlington    TX   
247                                       NaN    Wiliamsport    PA   
248                                       NaN  West New York    NJ   
249                                       NaN      Worcester    MA   
250                                       NaN      Worcester    MA   
251                                       NaN      Worcester    MA   

          start         end league  \
0    09/11/1880  05/30/1882     NL   
1    04/30/1884  05/31/1884     UA   
2    04/19/1966         NaN     AL   
3    04/21/1972  10/03/1993     AL   
4    04/11/1994         NaN     AL   
247  08/20/2017  08/20/2017     NL   
248  09/11/1898  09/17/1899     NL   
249  05/01/1880  09/29/1882     NL   
250  08/17/1887  08/17/1887     NL   
251  10/30/1874  10/30/1874    NaN   

                                                      notes  
0            TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1                                                       NaN  
2                                                       NaN  
3                                                       NaN  
4                                                       NaN  
247                                                     PIT  
248  BRO:9/18&10/2/1898; NY1:9/11/98, 6/4&7/16&8/13&9/17/99  
249                                                     NaN  
250                                              1 BSN game  
251                                              1 BS1 game

In [49]:
person_codes.head().append(person_codes.tail())

id      last   first player_debut mgr_debut coach_debut ump_debut
0      aardd001   Aardsma   David   04/06/2004       NaN         NaN       NaN
1      aaroh101     Aaron    Hank   04/13/1954       NaN         NaN       NaN
2      aarot101     Aaron  Tommie   04/10/1962       NaN  04/06/1979       NaN
3      aased001      Aase     Don   07/26/1977       NaN         NaN       NaN
4      abada001      Abad    Andy   09/10/2001       NaN         NaN       NaN
20489  zuvep001   Zuvella    Paul   09/04/1982       NaN  04/02/1996       NaN
20490  zuveg101  Zuverink  George   04/21/1951       NaN         NaN       NaN
20491  zwild101  Zwilling   Dutch   08/14/1910       NaN  04/15/1941       NaN
20492  zycht001      Zych    Tony   09/04/2015       NaN         NaN       NaN
20493  thoma102  Thompson     NaN          NaN       NaN         NaN       NaN

In [50]:
team_codes.head().append(team_codes.tail())

team_id league  start   end        city         nickname franch_id  seq
0       ALT     UA   1884  1884     Altoona  Mountain Cities       ALT    1
1       ARI     NL   1998     0     Arizona     Diamondbacks       ARI    1
2       BFN     NL   1879  1885     Buffalo           Bisons       BFN    1
3       BFP     PL   1890  1890     Buffalo           Bisons       BFP    1
4       BL1    NaN   1872  1874   Baltimore         Canaries       BL1    1
145     WS8     NL   1886  1889  Washington         Senators       WS8    1
146     WS9     AA   1891  1891  Washington         Senators       WS9    1
147     WSN     NL   1892  1899  Washington         Senators       WS9    2
148     WSU     UA   1884  1884  Washington        Nationals       WSU    1
149     MIA     NL   2012     0       Miami          Marlins       FLO    2

#### Description of the 4 tables:

<i><n> Game_log: This file has the statistics of each match played, including among others, date, home and away teams, referees_id, number of strakeouts, score, etc. There are 161 columns for 171,907 matches!

Park_code: Information about each Park or stadium. It has few columns such as name, name, id, state, city, name of the stadium, etc. There are 9 columns for 252 parks!

Person_codes: Information about each person who played professionally. Among others, name, surname, debut date, debut coach, etc. 7 columns by 20,494 people.

Team_codes: Team information such as nickname, id, city, start date, end date, franchise, league, etc. There are 9 columns for 150 teams.

#### DEFENSE POSITIONS (BASEBALL)
1: Pitcher, 2: Catcher, 3: First Base, 4: Second Base, 5: Short Stop, 6: Third Base, 7: Left Field, 8: Central Field, 9: Right Field. 

#### Information about the Leagues

NL: National League, founded: 1876, ceased: -
AL: American League, founded: 1901, ceased: -
AA: Double-A classification, founded: 1946, ceased: -
UA: Union Association, founded: 1984, ceased: 1984
FL: Florida State League, founded: 1919, ceased: - (minor league)
PL: Player League, founded: 1890, ceased: 1890. 

In [51]:
import sqlite3

def run_query(q):
    with sqlite3.connect('mlb.db') as conn:
        return pd.read_sql_query(q,conn)
    
def run_command(command):
    with sqlite3.connect('mlb.db') as conn:
       # conn.execute('PRAGMA foreign_keys = ON;')
        conn.isolation_level = None
        conn.execute(command)
        
def show_tables():
    query = 'SELECT name,type FROM sqlite_master \
            WHERE type IN ("table","view");'
    return run_query(query)
    
conn = sqlite3.connect('mlb.db')

run_command('DROP TABLE game_log')
run_command('DROP TABLE park_codes')
run_command('DROP TABLE person_codes')
run_command('DROP TABLE team_codes')
#run_command('DROP TABLE person_appearance')
run_command('DROP TABLE team')
run_command('DROP TABLE park')
run_command('DROP TABLE person')
run_command('DROP TABLE league')
run_command('DROP TABLE game')
run_command('DROP TABLE team_appearance')
run_command('DROP TABLE appearance_typepd')
run_command('DROP TABLE appearance_type')

In [52]:
game_log.to_sql('game_log',conn,index=False)
park_codes.to_sql('park_codes',conn,index=False)
person_codes.to_sql('person_codes',conn,index=False)
team_codes.to_sql('team_codes',conn,index=False)

run_command('ALTER TABLE game_log ADD COLUMN \
game_id NUMERIC')

In [53]:
show_tables()

name   type
0      game_log  table
1    park_codes  table
2  person_codes  table
3    team_codes  table

In [54]:
run_command('UPDATE game_log SET game_id = h_name||date||number_of_game;')

In [55]:
run_query('SELECT game_id FROM game_log LIMIT 5')

game_id
0  FW1187105040
1  WS3187105050
2  RC1187105060
3  CH1187105080
4  TRO187105090

###### NORMALIZATION:  

###### Repetition in columns: player columns in game_log.
cols 106 - 132: v_player_i_id, v_player_i_name, v_player_i_def_pos  i: 1...9
cols 133 - 159: h_player_i_id, h_player_i_name, h_player_i_def_pos  i: 1...9

In groups of 3, one player is described in batting order (1 to 9). "v" of visiting_starging, "h" of home_starting: we have player_id, player name and position in defense.

It could generate: a new table that groups the above and generates a list of players id that has its position in defense and its position in attack (would be the order of batting) and for each name assign a game_id (since for each game these positions may vary). This table should not have the names, we will have to put them in another table with id and name.


We could finally eliminate columns 106 to 159 including both, because for all this information, we have another table with 18 x 171.907 =
3,094,326 rows of 5 columns. The benefit is to remove 54 columns from game_log.

###### Non-primary keys should by atributes of a primary-key: Nombres de jugadores
A list of names with id could be generated.

You could remove the name of each player in game_log and just leave player_id.

###### Redundant data: park start and end info
dates of game_log can serve to derive the start and end of a park sice the game_log has the park_id and all dates of games.

##### Opportunities of normalization: 

1- the following columns have redundant data : 

90-91     Visiting team manager ID and name
92-93     Home team manager ID and name
94-95     Winning pitcher ID and name
96-97     Losing pitcher ID and name
98-99     Saving pitcher ID and name--"","(none)" if none awarded
100-101   Game Winning RBI batter ID and name--"","(none)" if none
          awarded
102-103   Visiting starting pitcher ID and name

Sice they have ID and name, we could put them in separate tables with unique id. That 'll generate 7 new tables. 

2- In game_log there is a date column (column 1 in "yyyymmdd") and a day_of_week column (column 3), then we can derive the second from the first one. 

3- 78 - 89 columns have info of the umpire ID and names. But I think that to generate extra tables from that doesn't add any value.

4- columns 4 - 5 and 7 - 8 have the following information respectively: visiting team and league, home team and league. Maybe, there is a unique league for each team, maybe not. 

Finally, all the statistical columns in game_log should be there because each game has unique statistics to summarize.

#### See: MLB_DB_Schema.png

### NORMALIZATION DEFINITIVE
(see MLB_DB_Schema.png)

#### Description and decisions taken on the following tables:
##### person
<i>Each of the 'debut' columns have been omitted, because all of the data will be able to be found from other tables.
Since the game log file has no data on coaches, we made the decision to not include this data.
##### park
The start, end, and league columns contain data that is found in the main game log and can be removed.
##### league
<i>Because some of the older leagues are not well known, we will create a table to store league names.
##### appearance_type
<i> Our appearance table will include data on players with positions, umpires, managers, and awards (like winning pitcher). This table will store information on what different types of appearances are available.
##### team
The start, end, and sequence columns can be derived from the game level data.
##### game
We have chosen to include all columns for the game log that don't refer to one specific team or player, instead we put those in two appearance tables.
We have removed the column with the day of the week, as this can be derived from the date.
We have changed the day_night column to day, with the intention of making this a boolean column. Even though SQLite doesn't support the BOOLEAN type, we can use this when creating our table and SQLite will manage the underlying types behind the scenes (for more on how this works refer to the SQLite documentation. This means that anyone quering the schema of our database in the future understands how that column is intended to be used.
##### team_appearance
The team_appearance table has a compound primary key composed of the team name and the game ID. In addition, a boolean column home is used to differentiate between the home and the away team. The rest of the columns are scores or statistics that in our original game log are repeated for each of the home and away teams.

In [56]:
query_person = 'CREATE TABLE IF NOT EXISTS person \
            (person_id TEXT PRIMARY KEY, \
            first_name TEXT, last_name TEXT);'

query_insert_data_person = 'INSERT INTO person \
            SELECT id, first, last FROM person_codes;'

run_command(query_person)
run_command(query_insert_data_person)

run_query('SELECT * FROM person LIMIT 5;')

person_id first_name last_name
0  aardd001      David   Aardsma
1  aaroh101       Hank     Aaron
2  aarot101     Tommie     Aaron
3  aased001        Don      Aase
4  abada001       Andy      Abad

In [57]:
query_park = 'CREATE TABLE IF NOT EXISTS park \
                (park_id TEXT PRIMARY KEY, \
                name TEXT, nickname TEXT, city TEXT, \
                state TEXT, notes TEXT);'

query_insert_data_park = 'INSERT INTO park \
            SELECT park_id, name, aka, city, state, notes \
            FROM park_codes;'

run_command(query_park)
run_command(query_insert_data_park)

run_query('SELECT * FROM park LIMIT 5;')

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                   nickname       city state  \
0                                      None     Albany    NY   
1                                      None    Altoona    PA   
2             Edison Field; Anaheim Stadium    Anaheim    CA   
3                                      None  Arlington    TX   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX   

                                            notes  
0  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1                                            None  
2                                            None  
3                                            None  
4                                            None

In [58]:
query_league = 'CREATE TABLE IF NOT EXISTS league \
                (league_id TEXT PRIMARY KEY, \
                name TEXT);'

query_insert_data_league = 'INSERT INTO league \
            VALUES ("NL", "National League"), \
                    ("AL", "American League"), \
                    ("AA", "Double-A classification"), \
                    ("UA", "Union Association"), \
                    ("FL", "Florida State League"), \
                    ("PL", "Player League");'

run_command(query_league)
run_command(query_insert_data_league)

run_query('SELECT * FROM league LIMIT 5;')

league_id                     name
0        NL          National League
1        AL          American League
2        AA  Double-A classification
3        UA        Union Association
4        FL     Florida State League

In [59]:
appearance_typepd = pd.read_csv('appearance_type.csv')
appearance_typepd

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

In [60]:
appearance_typepd.to_sql('appearance_typepd',conn,index=False)
run_query('SELECT * FROM appearance_typepd')

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

In [61]:
query_appearance_type = 'CREATE TABLE IF NOT EXISTS appearance_type \
                (appearance_type_id TEXT PRIMARY KEY, \
                name TEXT, category TEXT);'

query_insert_data_appearance_type = 'INSERT INTO appearance_type \
            SELECT * FROM appearance_typepd;'

run_command(query_appearance_type)
run_command(query_insert_data_appearance_type)

run_query('SELECT * FROM appearance_type;')

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

In [62]:
# we are going to eliminate row 113 from the team_codes table, because it repeats
# team_id (ML twice with league_id distinctions)
run_command('UPDATE team_codes SET team_id = "MIL2" \
            WHERE (team_id = "MIL") AND (league="NL")')

def run_command(command):
    with sqlite3.connect('mlb.db') as conn:
        conn.execute('PRAGMA foreign_keys = ON;')
        conn.isolation_level = None
        conn.execute(command)


query_team = 'CREATE TABLE IF NOT EXISTS team \
                (team_id TEXT PRIMARY KEY, \
                league_id TEXT, city TEXT, \
                nickname TEXT, franch_id TEXT, \
                FOREIGN KEY (league_id) REFERENCES league(league_id));'

query_insert_data_team = 'INSERT INTO team \
            SELECT team_id, league, city, nickname, \
            franch_id FROM team_codes;'

run_command(query_team)
run_command(query_insert_data_team)

run_query('SELECT * FROM team LIMIT 5;')

team_id league_id       city         nickname franch_id
0     ALT        UA    Altoona  Mountain Cities       ALT
1     ARI        NL    Arizona     Diamondbacks       ARI
2     BFN        NL    Buffalo           Bisons       BFN
3     BFP        PL    Buffalo           Bisons       BFP
4     BL1      None  Baltimore         Canaries       BL1

In [63]:
team_codes.loc[team_codes["team_id"] == 'MIL']
team_codes["team_id"].value_counts()

MIL    2
BR1    1
FW1    1
LS1    1
SLN    1
SL1    1
CIN    1
ELI    1
PHA    1
TL2    1
CN1    1
WIL    1
NYN    1
TRO    1
RC1    1
CH1    1
WS7    1
PIT    1
SL4    1
WS5    1
BRP    1
IND    1
SR2    1
WS1    1
BLN    1
NYA    1
CHU    1
PH1    1
OAK    1
BFP    1
COL    1
HR1    1
IN2    1
KC2    1
WSU    1
CH2    1
SDN    1
WSN    1
NY4    1
RC2    1
CL6    1
PHP    1
MIA    1
MLN    1
KEO    1
WAS    1
SL5    1
MLU    1
WS9    1
CHN    1
KCA    1
KCN    1
SPU    1
PHI    1
BSN    1
IN1    1
ML2    1
CLE    1
CN3    1
BRO    1
SEA    1
MLA    1
KCF    1
CHA    1
TEX    1
WS6    1
BLF    1
BFN    1
NEW    1
RIC    1
SLF    1
TOR    1
WS4    1
SE1    1
LS2    1
PTP    1
SFN    1
SR1    1
PT1    1
BLA    1
BRF    1
CL4    1
BLU    1
PHN    1
BUF    1
CL2    1
ARI    1
CNU    1
IN3    1
WS3    1
DTN    1
CHF    1
SL2    1
MID    1
CN2    1
HOU    1
BR2    1
NH1    1
CAL    1
BL2    1
BL4    1
BR4    1
SLU    1
KC1    1
TL1    1
SL3    1
BOS    1
CLP    1
HAR    1
TRN    1
BS1    1
L

In [64]:
query_game = 'CREATE TABLE IF NOT EXISTS game \
            (game_id TEXT PRIMARY KEY, \
            date INTEGER, number_of_game INTEGER, \
            park_id TEXT, length_outs REAL, day TEXT, \
            completion  TEXT, forefeit TEXT, \
            protest TEXT, attendance REAL, length_minutes REAL, \
            additional_info TEXT, acquisition_info TEXT, \
            FOREIGN KEY (park_id) REFERENCES park(park_id));'

query_data_game = 'INSERT INTO game \
                    SELECT game_id, date, number_of_game, \
                    park_id, length_outs, day_night, \
                    completion, forefeit, protest, attendance, \
                    length_minutes, additional_info, acquisition_info \
                    FROM game_log;'

run_command(query_game)
run_command(query_data_game)

run_query('SELECT * FROM game LIMIT 5')

game_id      date  number_of_game park_id  length_outs day completion  \
0  FW1187105040  18710504               0   FOR01         54.0   D       None   
1  WS3187105050  18710505               0   WAS01         54.0   D       None   
2  RC1187105060  18710506               0   RCK01         54.0   D       None   
3  CH1187105080  18710508               0   CHI01         54.0   D       None   
4  TRO187105090  18710509               0   TRO01         54.0   D       None   

  forefeit protest  attendance  length_minutes additional_info  \
0     None    None       200.0           120.0            None   
1     None    None      5000.0           145.0            HTBF   
2     None    None      1000.0           140.0            None   
3     None    None      5000.0           150.0            None   
4     None    None      3250.0           145.0            HTBF   

  acquisition_info  
0                Y  
1                Y  
2                Y  
3                Y  
4                Y

In [65]:
run_query('SELECT sql FROM sqlite_master WHERE name = \
"game_log" AND type = "table";')

sql
0  CREATE TABLE "game_log" (\n"date" INTEGER,\n  "number_of_game" INTEGER,\n  "day_of_week" TEXT,\n  "v_name" TEXT,\n  "v_league" TEXT,\n  "v_game_number" INTEGER,\n  "h_name" TEXT,\n  "h_league" TEXT,\n  "h_game_number" INTEGER,\n  "v_score" INTEGER,\n  "h_score" INTEGER,\n  "length_outs" REAL,\n  "day_night" TEXT,\n  "completion" TEXT,\n  "forefeit" TEXT,\n  "protest" TEXT,\n  "park_id" TEXT,\n  "attendance" REAL,\n  "length_minutes" REAL,\n  "v_line_score" TEXT,\n  "h_line_score" TEXT,\n  "v_at_bats" REAL,\n  "v_hits" REAL,\n  "v_doubles" REAL,\n  "v_triples" REAL,\n  "v_homeruns" REAL,\n  "v_rbi" REAL,\n  "v_sacrifice_hits" REAL,\n  "v_sacrifice_flies" REAL,\n  "v_hit_by_pitch" REAL,\n  "v_walks" REAL,\n  "v_intentional_walks" REAL,\n  "v_strikeouts" REAL,\n  "v_stolen_bases" REAL,\n  "v_caught_stealing" REAL,\n  "v_grounded_into_double" REAL,\n  "v_first_catcher_interference" REAL,\n  "v_left_on_base" REAL,\n  "v_pitchers_used" REAL,\n  "v_individual_earned_runs" REAL,\n  "v_team_earned_runs" REAL,\n  "v_wild_pitches" REAL,\n  "v_balks" REAL,\n  "v_putouts" REAL,\n  "v_assists" REAL,\n  "v_errors" REAL,\n  "v_passed_balls" REAL,\n  "v_double_plays" REAL,\n  "v_triple_plays" REAL,\n  "h_at_bats" REAL,\n  "h_hits" REAL,\n  "h_doubles" REAL,\n  "h_triples" REAL,\n  "h_homeruns" REAL,\n  "h_rbi" REAL,\n  "h_sacrifice_hits" REAL,\n  "h_sacrifice_flies" REAL,\n  "h_hit_by_pitch" REAL,\n  "h_walks" REAL,\n  "h_intentional_walks" REAL,\n  "h_strikeouts" REAL,\n  "h_stolen_bases" REAL,\n  "h_caught_stealing" REAL,\n  "h_grounded_into_double" REAL,\n  "h_first_catcher_interference" REAL,\n  "h_left_on_base" REAL,\n  "h_pitchers_used" REAL,\n  "h_individual_earned_runs" REAL,\n  "h_team_earned_runs" REAL,\n  "h_wild_pitches" REAL,\n  "h_balks" REAL,\n  "h_putouts" REAL,\n  "h_assists" REAL,\n  "h_errors" REAL,\n  "h_passed_balls" REAL,\n  "h_double_plays" REAL,\n  "h_triple_plays" REAL,\n  "hp_umpire_id" TEXT,\n  "hp_umpire_name" TEXT,\n  "1b_umpire_id" TEXT,\n  "1b_umpire_name" TEXT,\n  "2b_umpire_id" TEXT,\n  "2b_umpire_name" TEXT,\n  "3b_umpire_id" TEXT,\n  "3b_umpire_name" TEXT,\n  "lf_umpire_id" TEXT,\n  "lf_umpire_name" TEXT,\n  "rf_umpire_id" TEXT,\n  "rf_umpire_name" TEXT,\n  "v_manager_id" TEXT,\n  "v_manager_name" TEXT,\n  "h_manager_id" TEXT,\n  "h_manager_name" TEXT,\n  "winning_pitcher_id" TEXT,\n  "winning_pitcher_name" TEXT,\n  "losing_pitcher_id" TEXT,\n  "losing_pitcher_name" TEXT,\n  "saving_pitcher_id" TEXT,\n  "saving_pitcher_name" TEXT,\n  "winning_rbi_batter_id" TEXT,\n  "winning_rbi_batter_id_name" TEXT,\n  "v_starting_pitcher_id" TEXT,\n  "v_starting_pitcher_name" TEXT,\n  "h_starting_pitcher_id" TEXT,\n  "h_starting_pitcher_name" TEXT,\n  "v_player_1_id" TEXT,\n  "v_player_1_name" TEXT,\n  "v_player_1_def_pos" REAL,\n  "v_player_2_id" TEXT,\n  "v_player_2_name" TEXT,\n  "v_player_2_def_pos" REAL,\n  "v_player_3_id" TEXT,\n  "v_player_3_name" TEXT,\n  "v_player_3_def_pos" REAL,\n  "v_player_4_id" TEXT,\n  "v_player_4_name" TEXT,\n  "v_player_4_def_pos" REAL,\n  "v_player_5_id" TEXT,\n  "v_player_5_name" TEXT,\n  "v_player_5_def_pos" REAL,\n  "v_player_6_id" TEXT,\n  "v_player_6_name" TEXT,\n  "v_player_6_def_pos" REAL,\n  "v_player_7_id" TEXT,\n  "v_player_7_name" TEXT,\n  "v_player_7_def_pos" REAL,\n  "v_player_8_id" TEXT,\n  "v_player_8_name" TEXT,\n  "v_player_8_def_pos" REAL,\n  "v_player_9_id" TEXT,\n  "v_player_9_name" TEXT,\n  "v_player_9_def_pos" REAL,\n  "h_player_1_id" TEXT,\n  "h_player_1_name" TEXT,\n  "h_player_1_def_pos" REAL,\n  "h_player_2_id" TEXT,\n  "h_player_2_name" TEXT,\n  "h_player_2_def_pos" REAL,\n  "h_player_3_id" TEXT,\n  "h_player_3_name" TEXT,\n  "h_player_3_def_pos" REAL,\n  "h_player_4_id" TEXT,\n  "h_player_4_name" TEXT,\n  "h_player_4_def_pos" REAL,\n  "h_player_5_id" TEXT,\n  "h_player_5_name" TEXT,\n  "h_player_5_def_pos" REAL,\n  "h_player_6_id" TEXT,\n  "h_player_6_name" TEXT,\n  "h_player_6_def_pos" REAL,\n  "h_player_7_id" TEXT,\n  "h_player_7_name" TEXT,\n  "h_playe

In [66]:
query_team_appearance = """CREATE TABLE IF NOT EXISTS team_appearance 
(team_id TEXT, game_id TEXT, home INTEGER, league_id TEXT, score INTEGER, 
line_score TEXT, at_bats REAL, hits REAL, 
doubles REAL, triples REAL, 
homeruns REAL, rbi REAL, sacrifice_hits REAL,  
sacrifice_flies REAL, hit_by_pitch REAL, 
walks REAL, intentional_walks REAL, 
strikeouts REAL, stolen_bases REAL,
caught_stealing REAL, grounded_into_double REAL, 
first_catcher_interference REAL, left_on_base REAL, 
pitchers_used REAL, individual_earned_runs REAL, 
team_earned_runs REAL, wild_pitches REAL, 
balks REAL, putouts REAL, assists REAL, errors REAL, 
passed_balls REAL, double_plays REAL, triple_plays REAL, 
PRIMARY KEY (team_id,game_id), 
FOREIGN KEY (team_id) REFERENCES team(team_id), 
FOREIGN KEY (game_id) REFERENCES game(game_id), 
FOREIGN KEY (league_id) REFERENCES league(league_id) 
);"""

query_data_team_appearance = """INSERT INTO team_appearance 
SELECT h_name, game_id, 1 AS home, h_league, 
h_score, 
h_line_score, h_at_bats, h_hits, 
h_doubles, h_triples, 
h_homeruns, h_rbi, h_sacrifice_hits,  
h_sacrifice_flies, h_hit_by_pitch, 
h_walks, h_intentional_walks, 
h_strikeouts, h_stolen_bases, 
h_caught_stealing, h_grounded_into_double, 
h_first_catcher_interference, h_left_on_base, 
h_pitchers_used, h_individual_earned_runs, 
h_team_earned_runs, h_wild_pitches, 
h_balks, h_putouts, h_assists, h_errors, 
h_passed_balls, h_double_plays, h_triple_plays 
FROM game_log 
UNION 
SELECT v_name, game_id, 0 AS home, v_league, 
v_score, 
v_line_score, v_at_bats, v_hits, 
v_doubles, v_triples, 
v_homeruns, v_rbi, v_sacrifice_hits,  
v_sacrifice_flies, v_hit_by_pitch, 
v_walks, v_intentional_walks, 
v_strikeouts, v_stolen_bases, 
v_caught_stealing, v_grounded_into_double, 
v_first_catcher_interference, v_left_on_base, 
v_pitchers_used, v_individual_earned_runs, 
v_team_earned_runs, v_wild_pitches, 
v_balks, v_putouts, v_assists, v_errors, 
v_passed_balls, v_double_plays, v_triple_plays 
FROM game_log;"""

run_command(query_team_appearance)
run_command(query_data_team_appearance)

run_query('SELECT * FROM team_appearance LIMIT 5')

team_id       game_id  home league_id  score line_score at_bats  hits  \
0     ALT  ALT188404300     1        UA      2       None    None  None   
1     ALT  ALT188405020     1        UA      3       None    None  None   
2     ALT  ALT188405030     1        UA      5       None    None  None   
3     ALT  ALT188405050     1        UA      2       None    None  None   
4     ALT  ALT188405100     1        UA      9       None    None  None   

  doubles triples homeruns   rbi sacrifice_hits sacrifice_flies hit_by_pitch  \
0    None    None     None  None           None            None         None   
1    None    None     None  None           None            None         None   
2    None    None     None  None           None            None         None   
3    None    None     None  None           None            None         None   
4    None    None     None  None           None            None         None   

  walks intentional_walks strikeouts stolen_bases caught_stealing  \
0  None              None       None         None            None   
1  None              None       None         None            None   
2  None              None       None         None            None   
3  None              None       None         None            None   
4  None              None       None         None            None   

  grounded_into_double first_catcher_interference left_on_base pitchers_used  \
0                 None                       None         None          None   
1                 None                       None         None          None   
2                 None                       None         None          None   
3                 None                       None         None          None   
4                 None                       None         None          None   

  individual_earned_runs team_earned_runs wild_pitches balks putouts assists  \
0                   None             None         None  None    None    None   
1                   None             None         None  None    None    None   
2                   None             None         None  None    None    None   
3                   None             None         None  None    None    None   
4                   None             None         None  None    None    None   

  errors passed_balls double_plays triple_plays  
0   None         None         None         None  
1   None         None         None         None  
2   None         None         None         None  
3   None         None         None         None  
4   None         None         None         None

In [67]:
run_query('SELECT * FROM appearance_type')

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

In [68]:
query_person_appearance = """ CREATE TABLE IF NOT EXISTS 
person_appearance (appearance_id INTEGER PRIMARY KEY, 
person_id TEXT, team_id TEXT, game_id TEXT, 
appearance_type_id TEXT,
FOREIGN KEY (appearance_type_id) REFERENCES appearance_type(appearance_type_id),
FOREIGN KEY (game_id) REFERENCES game(game_id),
FOREIGN KEY (team_id) REFERENCES team(team_id),
FOREIGN KEY (person_id) REFERENCES person(person_id));"""

query_data_person_appearance = """INSERT INTO person_appearance
(game_id, team_id, person_id, appearance_type_id) 
SELECT game_id, NULL, hp_umpire_id, "UHP" FROM game_log WHERE hp_umpire_id IS NOT NULL
UNION
SELECT game_id, NULL, [1b_umpire_id], "U1B" FROM game_log WHERE [1b_umpire_id] IS NOT NULL
UNION
SELECT game_id, NULL, [2b_umpire_id], "U2B" FROM game_log WHERE [2b_umpire_id] IS NOT NULL
UNION
SELECT game_id, NULL, [3b_umpire_id], "U3B" FROM game_log WHERE [3b_umpire_id] IS NOT NULL
UNION
SELECT game_id, NULL, lf_umpire_id, "ULF" FROM game_log WHERE lf_umpire_id IS NOT NULL
UNION
SELECT game_id, NULL, rf_umpire_id, "URF" FROM game_log WHERE rf_umpire_id IS NOT NULL
UNION
SELECT game_id, v_name, v_manager_id, "MM" FROM game_log WHERE v_manager_id IS NOT NULL
UNION
SELECT game_id, h_name, h_manager_id, "MM" FROM game_log WHERE h_manager_id IS NOT NULL
UNION
SELECT game_id, NULL, winning_pitcher_id, "AWP" FROM game_log WHERE winning_pitcher_id IS NOT NULL
UNION
SELECT game_id, NULL, losing_pitcher_id, "ALP" FROM game_log WHERE losing_pitcher_id IS NOT NULL
UNION
SELECT game_id, NULL, saving_pitcher_id, "ASP" FROM game_log WHERE saving_pitcher_id IS NOT NULL
UNION
SELECT game_id, NULL, winning_rbi_batter_id, "AWB" FROM game_log WHERE winning_rbi_batter_id IS NOT NULL
UNION
SELECT game_id, v_name, v_starting_pitcher_id, "PSP" FROM game_log WHERE v_starting_pitcher_id IS NOT NULL
UNION
SELECT game_id, h_name, h_starting_pitcher_id, "PSP" FROM game_log WHERE h_starting_pitcher_id IS NOT NULL
;"""

query_data_person_appearance2 = """
INSERT INTO person_appearance (game_id, team_id, person_id, appearance_type_id)
SELECT
    game_id,
    {hv}_name,
    {hv}_player_{num}_id,
    "O{num}"
    FROM game_log WHERE {hv}_player_{num}_id IS NOT NULL
UNION
SELECT
    game_id,
    {hv}_name,
    {hv}_player_{num}_id,
    "D"||CAST({hv}_player_{num}_def_pos as INT)
    FROM game_log WHERE {hv}_player_{num}_id IS NOT NULL;
"""
run_command(query_person_appearance)
run_command(query_data_person_appearance)

for hv in ["h","v"]:
    for num in range(1,10):
        query_vars = {"hv":hv,"num":num}
        run_command(query_data_person_appearance2.format(**query_vars))

In [69]:
run_query('SELECT * FROM person_appearance LIMIT 100')

appearance_id person_id team_id       game_id appearance_type_id
0               1  maplb901    None  ALT188404300                UHP
1               2  curte801     ALT  ALT188404300                 MM
2               3  murpj104     ALT  ALT188404300                PSP
3               4  hodnc101     SLU  ALT188404300                PSP
4               5  sullt101     SLU  ALT188404300                 MM
5               6  hoopm101    None  ALT188405020                UHP
6               7  curte801     ALT  ALT188405020                 MM
7               8  learj102     ALT  ALT188405020                PSP
8               9  sullt101     SLU  ALT188405020                 MM
9              10  taylb103     SLU  ALT188405020                PSP
10             11  hoopm101    None  ALT188405030                UHP
11             12  connj102     ALT  ALT188405030                PSP
12             13  curte801     ALT  ALT188405030                 MM
13             14  hodnc101     SLU  ALT188405030                PSP
14             15  sullt101     SLU  ALT188405030                 MM
15             16  hoopm101    None  ALT188405050                UHP
16             17  browj106     ALT  ALT188405050                PSP
17             18  curte801     ALT  ALT188405050                 MM
18             19  sullt101     SLU  ALT188405050                 MM
19             20  taylb103     SLU  ALT188405050                PSP
20             21  maplb901    None  ALT188405100                UHP
21             22  browj106     ALT  ALT188405100                PSP
22             23  curte801     ALT  ALT188405100                 MM
23             24  bondt101     BSU  ALT188405100                PSP
24             25  murnt101     BSU  ALT188405100                 MM
25             26  maplb901    None  ALT188405120                UHP
26             27  curte801     ALT  ALT188405120                 MM
27             28  murpj104     ALT  ALT188405120                PSP
28             29  bondt101     BSU  ALT188405120                PSP
29             30  murnt101     BSU  ALT188405120                 MM
30             31  devid901    None  ALT188405140                UHP
31             32  browj106     ALT  ALT188405140                PSP
32             33  curte801     ALT  ALT188405140                 MM
33             34  bakej102     PHU  ALT188405140                PSP
34             35  malof101     PHU  ALT188405140                 MM
35             36  devid901    None  ALT188405150                UHP
36             37  curte801     ALT  ALT188405150                 MM
37             38  murpj104     ALT  ALT188405150                PSP
38             39  malof101     PHU  ALT188405150                 MM
39             40  weavs101     PHU  ALT188405150                PSP
40             41  devid901    None  ALT188405160                UHP
41             42  browj106     ALT  ALT188405160                PSP
42             43  curte801     ALT  ALT188405160                 MM
43             44  bakej102     PHU  ALT188405160                PSP
44             45  malof101     PHU  ALT188405160                 MM
45             46  devid901    None  ALT188405170                UHP
46             47  curte801     ALT  ALT188405170                 MM
47             48  murpj104     ALT  ALT188405170                PSP
48             49  bakej102     PHU  ALT188405170                PSP
49             50  malof101     PHU  ALT188405170                 MM
50             51  sulld901    None  ALT188405210                UHP
51             52  browj106     ALT  ALT188405210                PSP
52             53  curte801     ALT  ALT188405210                 MM
53             54  scanm801     WSU  ALT188405210                 MM
54             55  wiseb101     WSU  ALT188405210                PSP
55             56  sulld901    None  ALT188405230                UHP
56             57  curte801     ALT  ALT188405230                 MM
57

In [70]:
run_query('SELECT * FROM person WHERE first_name = "Isaac" AND last_name = "Leroy";')

person_id first_name last_name
0  leroi901      Isaac     Leroy

In [71]:
show_tables()

name   type
0            game_log  table
1          park_codes  table
2        person_codes  table
3          team_codes  table
4              person  table
5                park  table
6              league  table
7   appearance_typepd  table
8     appearance_type  table
9                team  table
10               game  table
11    team_appearance  table
12  person_appearance  table

In [72]:
run_command('DROP TABLE game_log;')
run_command('DROP TABLE park_codes;')
run_command('DROP TABLE team_codes;')
run_command('DROP TABLE person_codes;')